connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Loggers.java

```diff
@@ -174,43 +175,50 @@ private synchronized List<org.apache.log4j.Logger> loggers(String namespace) {
     }
 
     // visible for testing
-    org.apache.log4j.Logger lookupLogger(String logger) {
+    org.apache.logging.log4j.Logger lookupLogger(String logger) {
         return LogManager.getLogger(logger);
     }
 
-    @SuppressWarnings("unchecked")
-    // visible for testing
-    Enumeration<org.apache.log4j.Logger> currentLoggers() {
-        return LogManager.getCurrentLoggers();
+    List<org.apache.logging.log4j.Logger> currentLoggers() {
+        LoggerContext context = (LoggerContext) LogManager.getContext(false);
+        Collection<LoggerConfig> loggerConfigs = context.getConfiguration().getLoggers().values();
+        Set<String> loggerNames = loggerConfigs.stream()
+            .map(LoggerConfig::getName)
+            .collect(Collectors.toSet());
+
+        List<org.apache.logging.log4j.Logger> loggers = new ArrayList<>();
+        for (String name : loggerNames) {
+            loggers.add(LogManager.getLogger(name));
+        }
+        return loggers;
     }
 
     // visible for testing
-    org.apache.log4j.Logger rootLogger() {
+    org.apache.logging.log4j.Logger rootLogger() {
         return LogManager.getRootLogger();
     }
 
-    private void setLevel(org.apache.log4j.Logger logger, Level level) {
-        Level currentLevel = logger.getLevel();
-        if (currentLevel == null)
-            currentLevel = logger.getEffectiveLevel();
+    private void setLevel(org.apache.logging.log4j.Logger logger, Level level) {
+        String loggerName = logger.getName();
+        LoggerContext context = (LoggerContext) LogManager.getContext(false);
+        LoggerConfig loggerConfig = context.getConfiguration().getLoggerConfig(loggerName);
+        Level currentLevel = loggerConfig.getLevel();
```

Comment:
This looks pretty much as a maintenance headache for the Apache Kafka team. What will happen if the user switches logging implementation (at least 3 logging implementations are supported by the Log4j API, see [logging implementations](https://logging.apache.org/log4j/2.x/manual/installation.html#impl))?

It looks to me that you only use this for JMX. If that is the case, Log4j Core provides an [out-of-the-box JMX support](https://logging.apache.org/log4j/2.x/manual/jmx.html). You just need to enable it, since JMX is a potential source of security problems and is disabled by default.

If you need to get and set the levels for other reasons, please open a thread on `dev@logging`. Users like to change logger levels programmatically so often, that we'd better offer an implementation independent API for that.

Reply:
This is used by Kafka Connect. We have a REST API that allows changing the log level of all instances in a Kafka Connect cluster. See https://cwiki.apache.org/confluence/display/KAFKA/KIP-495%3A+Dynamically+Adjust+Log+Levels+in+Connect for the details.

Reply:
We can probably reach a consensus in the Logging PMC to release a new Log4j Configuration API, that you can use to abstract from the internals of the logging implementation (see [this thread on `dev@logging`](https://lists.apache.org/thread/ktdlo5br6jc39xo33vdsfsj2s200mm93)).

What is the planned release date for Kafka 4.x? If you wait until the end of the year, this class might not be necessary.

Reply:
Hi @ppkarwasz 
Currently, AK 4.0 release is scheduled at January 29th 2025.
For further details, you can refer to the release plan:
https://cwiki.apache.org/confluence/display/KAFKA/Release+Plan+4.0.0